web/src/message_list.js

```diff
@@ -335,22 +335,22 @@ export class MessageList {
         opts.id = id;
         this.data.set_selected_id(id);
 
+        // Avoid recursive calls to message_selected event
+        // by temporarily blocking triggering the event by
+        // setting `should_trigger_message_selected_event`.
         if (opts.force_rerender) {
-            // TODO: Because rerender() itself will call
-            // reselect_selected_id after doing the rendering, we
-            // actually end up with this function being called
-            // recursively in this case. The ordering will end up
-            // being that the message_selected.zulip event for that
-            // rerender is processed before execution returns here.
-            //
-            // The recursive call is unnecessary, so we should figure
-            // out how to avoid that, both here and in the next block.
+            this.should_trigger_message_selected_event = false;
             this.rerender();
+            this.should_trigger_message_selected_event = true;
         } else if (!opts.from_rendering) {
+            this.should_trigger_message_selected_event = false;
             this.view.maybe_rerender();
+            this.should_trigger_message_selected_event = true;
```

Comment:
Would it be safer to use a `try/finally` block here? I guess it's unlikely to leak. 

I also wonder if one could do an early return instead of the if statement below? I assume the recursive calls won't match either `!from_rendering` or `force_rerender`, and it might be more obviously correct to not have to think about if one of these might be true in that recursive call and thus set `this.should_trigger_message_selected_event = true`.

And then it might look more like passing a parameter into the rerender code path that's a `from_select_message`, rather than having global state? 

Not sure if this'd look nice, but maybe worth thinking about?

Reply:
Yeah, I tried thinking about different ways we can approach this which were the same as the ones stated by you above but this approach seemed like the most audiable and clear one to me. Also, if a `rerender` or `maybe_rerender` is setting `this.should_trigger_message_selected_event = true` then we are re-rendering twice without any additional information which is a bug.